I'm experimenting with an approach to improve how Spoken Dialog Systems (like OpenAI's Advanced Voice Mode) identify the end of a user's converational turn so the system can jump in appropriately without interrupting the user mid-thought.

As a user, current systems appear to rely primarily on pause length from a Voice Activity Detector (VAD). So an `X` second pause mid-sentence is treated similarly to `X` seconds of silence at the end of a sentence.

This notebook adds an end-of-turn prediction head to a Whisper model. This prediction head relies on both acoustic and linguistic information (both the Whisper encoder and the Whisper decoder). I will train it from multi-turn conversations between human speakers with labeled transitions. Implicitly, I assume humans do well at knowing when to jump in without interrupting ¯\_(ツ)_/¯

Candidate datasets for training: AMI Meeting Corpus, Switchboard, Fisher English Training Speech, DailyTalk, CALLHOME, CoVoST. If necessary, I could use diarazation tools like pyannotate-audio or pyAudioAnalysis to identify changes in the speaker.

A production system might still use a VAD as a computationally cheap approach to identify pauses of at least 1 second, and then I'd call my updated model from this notebook for final end-of-turn detection and transcription only when there is some meaningful pause.

This is an educational project to gain initial experience with audio models. I may be naively underestimating what others before me have done. I will nevertheless run incremental experiments as a learning experience. The main steps are:

- [x] Download a Whisper model from HuggingFace hub and verify that I can run it on a trivial file
- [ ] Inspect the architecture and plan how to integrate a prediction head
- [ ] Overfit on a single sample to verify I can train
- [ ] Train on a set of ~100 samples with some small number of validation samples to test infrastructure
- [ ] Train on a larger sample to test if I can make something that broadly "works"

# Example Calling A Whisper Model

I start with `whisper-tiny.en`. Later stages may use a large model like `whisper-large-v3` once I have the basic workflow wired up.

In [1]:
from time import time
import torch
from transformers import AutoModelForSpeechSeq2Seq, AutoProcessor, pipeline

device = "cuda:0" if torch.cuda.is_available() else "cpu"
torch_dtype = torch.float16 if torch.cuda.is_available() else torch.float32

model_id = "openai/whisper-tiny.en"

model = AutoModelForSpeechSeq2Seq.from_pretrained(
    model_id, torch_dtype=torch_dtype, low_cpu_mem_usage=True, use_safetensors=True
)
model.to(device)
 
processor = AutoProcessor.from_pretrained(model_id)

pipe = pipeline(
    "automatic-speech-recognition",
    model=model,
    tokenizer=processor.tokenizer,
    feature_extractor=processor.feature_extractor,
    torch_dtype=torch_dtype,
    device=device,
)

start = time()
result = pipe("./test_data/test.wav")
print(f"Type: {type(result)}")
print(f"Keys: {result.keys() if isinstance(result, dict) else 'not a dict'}")
print(result["text"])
print(f"Time taken: {time() - start}")
print(dir(pipe))


Device set to use cpu
/Users/dan/interwhisp/iw_venv/lib/python3.11/site-packages/transformers/models/whisper/generation_whisper.py:512: FutureWarning: The input name `inputs` is deprecated. Please make sure to use `input_features` instead.
  warnings.warn(


Type: <class 'dict'>
Keys: dict_keys(['text'])
 My name is Dan and this is a test audio file.
Time taken: 0.20616602897644043
['__abstractmethods__', '__call__', '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getstate__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__slots__', '__str__', '__subclasshook__', '__weakref__', '_abc_impl', '_batch_size', '_create_repo', '_ensure_tensor_on_device', '_forward', '_forward_params', '_get_files_timestamps', '_load_feature_extractor', '_load_image_processor', '_load_processor', '_load_tokenizer', '_num_workers', '_postprocess_params', '_preprocess_params', '_sanitize_parameters', '_upload_modified_files', 'binary_output', 'call_count', 'check_model_type', 'default_input_names', 'device', 'device_placement', 'ensure_tensor_on_device', 'feat

# Inspect the Pipeline and Model

In [2]:
# Whats' in the pipeline?
print([step for step in pipe.__dict__.keys() if not step.startswith("_")])


['type', 'task', 'model', 'tokenizer', 'feature_extractor', 'image_processor', 'processor', 'modelcard', 'framework', 'device', 'binary_output', 'prefix', 'generation_config', 'call_count']


In [3]:
pipe.feature_extractor

WhisperFeatureExtractor {
  "chunk_length": 30,
  "feature_extractor_type": "WhisperFeatureExtractor",
  "feature_size": 80,
  "hop_length": 160,
  "n_fft": 400,
  "n_samples": 480000,
  "nb_max_frames": 3000,
  "padding_side": "right",
  "padding_value": 0.0,
  "processor_class": "WhisperProcessor",
  "return_attention_mask": false,
  "sampling_rate": 16000
}

Pipeline expects incoming audio at 16kHz. Preprocessor creates 80 dimensional features in 0.01second increments (the hop length of 160 means 160 frames/samples of audio are combined into a single input frame to the model)

In [4]:
from torchinfo import summary
decoder_input_ids = torch.tensor([[processor.tokenizer.pad_token_id]]).to(device)
print(summary(pipe.model, input_size=(1, 80, 3000), decoder_input_ids=decoder_input_ids))

Passing a tuple of `past_key_values` is deprecated and will be removed in Transformers v4.43.0. You should pass an instance of `EncoderDecoderCache` instead, e.g. `past_key_values=EncoderDecoderCache.from_legacy_cache(past_key_values)`.


Layer (type:depth-idx)                                  Output Shape              Param #
WhisperForConditionalGeneration                         [1, 1500, 384]            --
├─WhisperModel: 1-1                                     [1, 1500, 384]            --
│    └─WhisperEncoder: 2-1                              [1, 1500, 384]            576,000
│    │    └─Conv1d: 3-1                                 [1, 384, 3000]            92,544
│    │    └─Conv1d: 3-2                                 [1, 384, 1500]            442,752
│    │    └─ModuleList: 3-3                             --                        7,096,320
│    │    └─LayerNorm: 3-4                              [1, 1500, 384]            768
│    └─WhisperDecoder: 2-2                              [1, 6, 1, 64]             --
│    │    └─Embedding: 3-5                              [1, 1, 384]               19,915,776
│    │    └─WhisperPositionalEmbedding: 3-6             [1, 1, 384]               172,032
│    │    └─ModuleList: 3

# Adding TurnEndClassifier to Whisper Model

I add a TurnEndClassifier as an additional prediction head for the Whisper model. It's inputs are 
1. audio features from the encoder
2. Semantic info from decoder

## Alignment

Each recording will be classified with a single prediction of whether it ends at a turn-end. The recording has many values from the encoder on the time dimension (100 samples per second) and many values from the decoder (1 hidden state per token).

I compress the time and token dimensions from the encoder and decoder respectively into a 1d representation from each. I do these compressions with convolutional layers and then pooling layers (separate layers for the encoder and the decoder compression to 1d). These 1d vectors are concatenated, fed through a small feedforward network, and lead to a binary classification head indicating if this audio finishes with a turn-end.

For training, I'll preprocess conversations to have many samples that are each 10 seconds of audio and that finish at a turn end (target is 1) or that don't (target is 0).

In [5]:
from torch import nn
import torch

class TurnEndClassifier(nn.Module):
    def __init__(self, encoder_dim=384, decoder_dim=384, hidden_dim=64):
        super(TurnEndClassifier, self).__init__()

        # Linear layers to reduce the dimensions of the encoder and decoder outputs.
        # This is a temporary hack. In theory, hidden_dim could be larger than whisper_hidden_dim. Or encoder and decoder could have different dimensions.
        self.encoder_reduce = nn.Linear(encoder_dim, hidden_dim)
        self.decoder_reduce = nn.Linear(decoder_dim, hidden_dim)
        
        # Encoder processing layers
        self.encoder_conv = nn.Sequential(
            nn.Conv1d(hidden_dim, hidden_dim, kernel_size=5, stride=3, padding=1),
            nn.ReLU(),
            nn.Conv1d(hidden_dim, hidden_dim, kernel_size=5, stride=3, padding=1),
            nn.ReLU(),
            nn.Conv1d(hidden_dim, hidden_dim, kernel_size=5, stride=3, padding=1),
            nn.ReLU()
        )
        self.encoder_pool = nn.AdaptiveMaxPool1d(1)
        
        # Decoder processing layers
        self.decoder_conv = nn.Sequential(
            nn.Conv1d(hidden_dim, hidden_dim, kernel_size=3, padding=1),
            nn.ReLU(),
            nn.Conv1d(hidden_dim, hidden_dim, kernel_size=3, padding=1),
            nn.ReLU()
        )
        self.decoder_pool = nn.AdaptiveMaxPool1d(1)
        
        # Fully connected layers for classification
        self.fc = nn.Sequential(
            nn.Linear(2*hidden_dim , hidden_dim),
            nn.ReLU(),
            nn.Linear(hidden_dim, 1),
            nn.Sigmoid()
        )
    
    def forward(self, encoder_outputs, decoder_outputs):
        print(f"TurnEndClassifier input shapes:")
        print(f"- encoder_outputs: {encoder_outputs.shape}")  # [batch, time, features]
        print(f"- decoder_outputs: {decoder_outputs.shape}")  # [batch, seq_len, features]

        # Reduce feature dimension first (more efficient for subsequent conv layers)
        # [batch, time, whisper_hidden_dim] -> [batch, time, hidden_dim]
        encoder_outputs = self.encoder_reduce(encoder_outputs)
        
        # [batch, seq_len, whisper_hidden_dim] -> [batch, seq_len, hidden_dim]
        decoder_outputs = self.decoder_reduce(decoder_outputs)
        
        # Process encoder outputs
        # [batch, time, hidden_dim] -> [batch, hidden_dim, time]
        encoder_outputs = encoder_outputs.transpose(1, 2)  # Conv1d expects channels first
        encoder_features = self.encoder_conv(encoder_outputs)  # Now working with reduced dimensions
        encoder_features = self.encoder_pool(encoder_features).squeeze(-1)  # -> [batch, hidden_dim]
        
        # Process decoder outputs
        # [batch, seq_len, hidden_dim] -> [batch, hidden_dim, seq_len]
        decoder_outputs = decoder_outputs.transpose(1, 2)  # Conv1d expects channels first
        decoder_features = self.decoder_conv(decoder_outputs)  # Now working with reduced dimensions
        decoder_features = self.decoder_pool(decoder_features).squeeze(-1)  # -> [batch, hidden_dim]
        
        # Concatenate and classify
        # [batch, hidden_dim] + [batch, hidden_dim] -> [batch, 2*hidden_dim]
        combined_features = torch.cat((encoder_features, decoder_features), dim=1)
        # [batch, 2*hidden_dim] -> [batch, 1]
        output = self.fc(combined_features)
        
        return output

## Custom Model
Now we build the custom model that is a Whisper Model with the extra prediction head. Our goals include:
1. Allow training the layers in the TurnEndClassifier prediction head while keeping all other layers frozen
2. Allow embedding this in a pipeline that is as similar as possible to the pipeline used at the top of this notebook.

In [6]:
from transformers import WhisperForConditionalGeneration
import torch


class CustomWhisperModel(WhisperForConditionalGeneration):
    def __init__(self, config, turn_end_classifier=None):
        super().__init__(config)
        self.turn_end_classifier = turn_end_classifier # if turn_end_classifier is None, we'll set it later

    @classmethod
    def from_pretrained(cls, pretrained_model_name_or_path, turn_end_classifier, *args, **kwargs):
        model = super().from_pretrained(pretrained_model_name_or_path, *args, **kwargs)
        model.turn_end_classifier = turn_end_classifier
        return model

    def forward_with_turn_end(
        self,
        input_features=None,
        decoder_input_ids=None,
        **kwargs  # Catch all other args to pass through
    ):
        assert self.turn_end_classifier is not None, "TurnEndClassifier must be set before calling forward_with_turn_end"
        
        # whisper_outputs is a Seq2SeqLMOutput object with `logits` for each token, encoder_last_hidden_state, and decoder_hidden_states, decoder_hidden_states, encoder_hidden_states and some other attributes. It's a type of namedtuple.
        whisper_outputs = super().forward(
            input_features=input_features,
            decoder_input_ids=decoder_input_ids,
            output_hidden_states=True,
            **kwargs
        )

        # Get output of last hidden layer from both encoder and decoder.
        encoder_hidden_states = whisper_outputs.encoder_last_hidden_state
        decoder_hidden_states = whisper_outputs.decoder_hidden_states[-1]

        # Get turn-end predictions
        turn_end_predictions = self.turn_end_classifier(
            encoder_hidden_states,
            decoder_hidden_states
        )
        
        return whisper_outputs, turn_end_predictions


## Test the CustomWhisperModel

In [7]:
def test_custom_whisper_model():
    # 1. Create fake data
    batch_size = 2
    sequence_length = 3000
    encoder_dim = 384  # Whisper tiny dimension
    decoder_seq_length = 20
    
    fake_input_features = torch.randn(batch_size, 80, sequence_length)
    fake_decoder_input_ids = torch.randint(0, 100, (batch_size, decoder_seq_length))
    
    # 2. Initialize models
    turn_end_classifier = TurnEndClassifier(
        encoder_dim=encoder_dim,
        decoder_dim=encoder_dim,
        hidden_dim=64
    )
    
    model = CustomWhisperModel.from_pretrained(
        "openai/whisper-tiny.en",
        turn_end_classifier=turn_end_classifier
    )
    
    # 3. Run forward pass

    whisper_outputs, turn_end_predictions = model.forward_with_turn_end(
        input_features=fake_input_features,
        decoder_input_ids=fake_decoder_input_ids,
    )
    print(f"Output type: {type(whisper_outputs)}")
    print("\nAvailable attributes:")
    print([attr for attr in dir(whisper_outputs) if not attr.startswith('_')])
    
    print("\nKey shapes:")
    print(f"- logits for word probabilities: {whisper_outputs.logits.shape}")
    print(f"- turn_end_predictions: {turn_end_predictions}")
    
        
test_custom_whisper_model()

Initializing TurnEndClassifier with dims: encoder=384, decoder=384
TurnEndClassifier input shapes:
- encoder_outputs: torch.Size([2, 1500, 384])
- decoder_outputs: torch.Size([2, 20, 384])
Output type: <class 'transformers.modeling_outputs.Seq2SeqLMOutput'>

Available attributes:
['clear', 'copy', 'cross_attentions', 'decoder_attentions', 'decoder_hidden_states', 'encoder_attentions', 'encoder_hidden_states', 'encoder_last_hidden_state', 'fromkeys', 'get', 'items', 'keys', 'logits', 'loss', 'move_to_end', 'past_key_values', 'pop', 'popitem', 'setdefault', 'to_tuple', 'update', 'values']

Key shapes:
- logits for word probabilities: torch.Size([2, 20, 51864])
- turn_end_predictions: tensor([[0.5481],
        [0.5333]], grad_fn=<SigmoidBackward0>)


# Wrap Custom Model In A Custom Pipeline

Create a pipeline that can take a .wav file as input and return both the text and turn_end_probability. After this is working, I will later explore adding functionality to run this without doing the sampling of tokens. Sampling is autoregressive, so it may be slow. And could theoretically run just end-of-turn classification, and then get the transcription iff we find it is in fact the end of turn.

In [13]:
from transformers import AutomaticSpeechRecognitionPipeline
from dataclasses import dataclass
from typing import Dict, Optional
import torch
from enum import Enum

@dataclass
class TurnEndContext:
    """Context object holding state between turn-end prediction and transcription"""
    encoder_outputs: torch.Tensor
    decoder_outputs: torch.Tensor
    attention_mask: Optional[torch.Tensor]

class TurnEndPipeline(AutomaticSpeechRecognitionPipeline):
    def __call__(self, *args, **kwargs):
        raise ValueError(
            "This pipeline uses a two-stage approach. Please use predict_turn_end() "
            "followed by transcribe() instead of calling the pipeline directly"
        )
    def predict_turn_end(self, audio_input):
        """First stage: Predict if audio ends at a turn boundary."""
        model_inputs_generator = self.preprocess(audio_input)
        
        results = []
        contexts = []
        
        # Process each chunk/batch item
        for model_inputs in model_inputs_generator:
            # Get turn-end prediction and cache states
            model_kwargs = {
                "input_features": model_inputs["input_features"],
                "decoder_input_ids": torch.tensor([[self.model.config.decoder_start_token_id]]).to(self.device)
            }
            if "attention_mask" in model_inputs:
                model_kwargs["attention_mask"] = model_inputs["attention_mask"]

            # Forward pass through our custom Whisper model
            whisper_outputs, turn_end_predictions = self.model.forward_with_turn_end(**model_kwargs)
            
            # Create context object for potential transcription
            contexts.append(TurnEndContext(
                encoder_outputs=whisper_outputs.encoder_last_hidden_state,
                decoder_outputs=whisper_outputs.decoder_hidden_states[-1],
                attention_mask=model_inputs.get("attention_mask", None)
            ))
            
            results.append({
                "turn_end_probability": turn_end_predictions.squeeze().item()
            })
        
        # For now, just return the first result/context
        # TODO: Handle multiple chunks/batches appropriately
        return results[0], contexts[0]


    def transcribe(self, context: TurnEndContext):
        """Second stage: Generate transcription using context from first stage."""
        if not isinstance(context, TurnEndContext):
            raise ValueError(
                "Transcribe requires a TurnEndContext object from predict_turn_end(). "
                "Please call predict_turn_end() first."
            )
            
        # Use cached states to generate transcription
        generated_ids = self.model.generate(
            encoder_outputs=context.encoder_outputs,
            attention_mask=context.attention_mask
        )
        
        # Decode the generated IDs to text
        text = self.tokenizer.batch_decode(generated_ids, skip_special_tokens=True)[0].strip()
        
        return {"text": text}
# Create the turn end classifier
turn_end_classifier = TurnEndClassifier(
    encoder_dim=384,  # Whisper tiny dimension
    decoder_dim=384,
    hidden_dim=64
).to(device)

# Create the custom model
model = CustomWhisperModel.from_pretrained(
    model_id,
    turn_end_classifier=turn_end_classifier,
    torch_dtype=torch_dtype,
    low_cpu_mem_usage=True,
    use_safetensors=True
).to(device)

# Create the processor
processor = AutoProcessor.from_pretrained(model_id)

pipe = TurnEndPipeline(
    model=model,
    tokenizer=processor.tokenizer,
    feature_extractor=processor.feature_extractor,
    torch_dtype=torch_dtype,
    device=device,
)

# Two-stage usage
result, context = pipe.predict_turn_end("./test_data/test.wav")
print(f"Turn end probability: {result['turn_end_probability']:.2f}")

if result['turn_end_probability'] > 0.5:  # or whatever threshold you choose
    transcription = pipe.transcribe(context)
    print(f"Transcription: {transcription['text']}")

Initializing TurnEndClassifier with dims: encoder=384, decoder=384


Device set to use cpu


TypeError: 'generator' object is not subscriptable